execution/executor/src/components/block_tree/mod.rs

```diff
@@ -0,0 +1,257 @@
+                        weak.upgrade()
+                            .ok_or_else(|| anyhow!("Block {:x} has been deallocated.", id))
+                    })
+                    .transpose()
+            })
+            .collect()
+    }
+
+    fn get(&self, id: HashValue) -> Result<Option<Arc<Block>>> {
+        Ok(self.multi_get(&[id])?.pop().expect("Must exist."))
+    }
+}
+
+struct BlockLookup {
+    inner: Mutex<Inner>,
+}
+
+impl BlockLookup {
+    fn new() -> Self {
+        Self {
+            inner: Mutex::new(Inner(HashMap::new())),
+        }
+    }
+
+    fn multi_get(&self, ids: &[HashValue]) -> Result<Vec<Option<Arc<Block>>>> {
+        self.inner.lock().multi_get(ids)
+    }
+
+    fn fetch_or_add_block(
+        self: &Arc<Self>,
+        id: HashValue,
+        output: ExecutedChunk,
+        parent_id: Option<HashValue>,
+    ) -> Result<Arc<Block>> {
+        let (block, existing, parent_block) = {
+            let mut locked = self.inner.lock();
+            let parent_block = parent_id
+                .map(|id| {
+                    locked
+                        .get(id)?
+                        .ok_or_else(|| anyhow!("parent block {} doesn't exist."))
+                })
+                .transpose()?;
+
+            match locked.0.entry(id) {
+                Entry::Occupied(entry) => {
+                    let existing = entry
+                        .get()
+                        .upgrade()
+                        .ok_or_else(|| anyhow!("block dropped unexpected."))?;
+                    ensure!(
+                        existing
+                            .output
+                            .result_view
+                            .is_same_view(&output.result_view),
+                        "Different block with same id {}",
+                        id,
+                    );
+                    (existing, true, parent_block)
+                }
+                Entry::Vacant(entry) => {
+                    let block = Arc::new(Block {
+                        id,
+                        output,
+                        children: Mutex::new(Vec::new()),
+                        block_lookup: self.clone(),
+                    });
+                    entry.insert(Arc::downgrade(&block));
+                    (block, false, parent_block)
+                }
+            }
+        };
+
+        if let Some(parent_block) = parent_block {
+            if existing {
+                parent_block.ensure_has_child(id)?;
+            } else {
+                parent_block.add_child(block.clone());
+            }
+        }
+
+        Ok(block)
+    }
+
+    fn remove(&self, id: HashValue) {
+        self.inner.lock().0.remove(&id);
+    }
+}
+
+pub struct BlockTree {
+    root: Mutex<Arc<Block>>,
+    block_lookup: Arc<BlockLookup>,
+}
+
+impl BlockTree {
+    pub fn new(db: &Arc<dyn DbReader<DpnProto>>) -> Result<Self> {
+        let block_lookup = Arc::new(BlockLookup::new());
```

Comment:
i'd call it `id_to_block`

Reply:
the type as well or only the field in the BlockTree structure?

Reply:
probably struct as well? the inner name is also confusing, maybe

Reply:
`Block::id_to_block` <- especially weird to me
`BlockTree::id_to_block`
and `struct IdToBlock`?

`Inner` is commonly used I think?

Reply:
I'd just use id_to_block: HashMap<_,_> instead of two wrappers. I also don't think we need the fancy GC-like behavior. we just use explicit prune in https://github.com/diem/diem/blob/1b0fc1d9d391f2bf59e03b3d60f570acbc7f4290/consensus/src/block_storage/block_tree.rs#L320

but I don't care much about it assuming it'd be merged back to consensus soon.

Reply:
>  I don't care much about it assuming it'd be merged back to consensus soon.
We are on the same page.

>  I also don't think we need the fancy GC-like behavior. we just use explicit prune in
Nah, nah, I like to be fancy and I rock 🤟